File: src/api_routes.rs
use std::collections::BTreeMap;
use std::result::Result;

use std::time::SystemTime;
use std::sync::Arc;

use rocket::{Route, State};

use rocket::http::{Cookie, CookieJar};


use rocket::response::{Redirect};
use rocket::serde::json::{Json, Value};
use serde_json::json;


use crate::helpers::{before};
use crate::my_types::{MonoAttrs, MyMods, Config, CfgAndLU, SgroupAndMoreOut, RemoteConfig, Right, Subjects, Mright, SgroupsWithAttrs, SubjectsAndCount, LdapConfigOut};
use crate::api_get;
use crate::api_post;
use crate::rocket_helpers::{OrigUrl, MyJson, action_result, to_json, err_to_json, Cache};
use crate::test_data;
use crate::cas_auth;


#[get("/login?<target>&<ticket>")]
async fn login(target: String, ticket: String, orig_url: OrigUrl, cookie_jar: &CookieJar<'_>, config: &State<Config>) -> Result<Redirect, String> {
    if !target.starts_with('/') || target.starts_with("//") {
        return Err(format!("invalid target {}, it must be a path-absolute url", target));
    }
    let service = before(&orig_url.0, "&ticket=").ok_or("weird login url")?;
    let user = cas_auth::validate_ticket(&config.cas.prefix_url, service, &ticket).await?;
    cookie_jar.add_private(Cookie::new("user_id", user));
    Ok(Redirect::found(target))
}

#[get("/set_test_data")]
async fn set_test_data(cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(test_data::set(cfg_and_lu).await)
}
#[get("/clear_test_data")]
async fn clear_test_data(cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(test_data::clear(&cfg_and_lu).await)
}
#[get("/add_test_data")]
async fn add_test_data(cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(test_data::add(cfg_and_lu).await)
}

#[post("/create?<id>", data = "<attrs>")]
async fn create(id: String, attrs: Json<MonoAttrs>, cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(
        api_post::create(cfg_and_lu, &id, attrs.into_inner()).await
    )
}

#[post("/modify_sgroup_attrs?<id>", data = "<attrs>")]
async fn modify_sgroup_attrs(id: String, attrs: Json<MonoAttrs>, cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(
        api_post::modify_sgroup_attrs(cfg_and_lu, &id, attrs.into_inner()).await
    )
}

#[post("/delete?<id>")]
async fn delete(id: String, cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(api_post::delete(cfg_and_lu, &id).await)
}

#[post("/modify_members_or_rights?<id>", data = "<mods>")]
async fn modify_members_or_rights(id: String, mods: Json<MyMods>, cfg_and_lu : CfgAndLU<'_>) -> MyJson {
    action_result(api_post::modify_members_or_rights(cfg_and_lu, &id, mods.into_inner()).await)
}

#[get("/sgroup?<id>")]
async fn sgroup(id: String, cfg_and_lu : CfgAndLU<'_>) -> Result<Json<SgroupAndMoreOut>, MyJson> {
    to_json(api_get::get_sgroup(cfg_and_lu, &id).await)
}

#[get("/sgroup_direct_rights?<id>")]
async fn sgroup_direct_rights(id: String, cfg_and_lu : CfgAndLU<'_>) -> Result<Json<BTreeMap<Right, Subjects>>, MyJson> {
    to_json(api_get::get_sgroup_direct_rights(cfg_and_lu, &id).await)
}

#[get("/group_flattened_mright?<id>&<mright>&<search_token>&<sizelimit>")]
async fn group_flattened_mright(id: String, mright: String, search_token: Option<String>, sizelimit: Option<usize>, cfg_and_lu : CfgAndLU<'_>) -> Result<Json<SubjectsAndCount>, MyJson> {
    let mright = Mright::from_string(&mright).map_err(err_to_json)?;
    to_json(api_get::get_group_flattened_mright(cfg_and_lu, &id, mright, search_token, sizelimit).await)
}

#[get("/search_sgroups?<right>&<search_token>&<sizelimit>")]
async fn search_sgroups(right: String, search_token: String, sizelimit: i32, cfg_and_lu : CfgAndLU<'_>) -> Result<Json<SgroupsWithAttrs>, MyJson> {
    let right = Right::from_string(&right).map_err(err_to_json)?;
    to_json(api_get::search_sgroups(cfg_and_lu, right, search_token, sizelimit).await)
}

#[get("/mygroups")]
async fn mygroups(cfg_and_lu : CfgAndLU<'_>) -> Result<Json<SgroupsWithAttrs>, MyJson> {
    to_json(api_get::mygroups(cfg_and_lu).await)
}

#[get("/clear_cache")]
async fn clear_cache(cache : &State<Cache>) {
    let mut val = cache.synchronized_groups.lock().unwrap();
    *val = Some((SystemTime::now(), Arc::new(hashmap!["foo".to_owned() => "bar2".to_owned()])));
}

#[get("/search_subjects?<search_token>&<sizelimit>&<source_dn>")]
async fn search_subjects(search_token: String, sizelimit: i32, source_dn: Option<String>, cfg_and_lu : CfgAndLU<'_>) -> Result<Json<BTreeMap<&String /* ssdn */, Subjects>>, MyJson> {
    to_json(api_get::search_subjects(cfg_and_lu, search_token, sizelimit, source_dn).await)
}

#[get("/config/public")]
fn config_public(cfg : &State<Config>) -> Value {
    json!({
        "cas_prefix_url": cfg.cas.prefix_url,
    })
}
#[get("/config/subject_sources")]
fn config_subject_sources(cfg_and_lu : CfgAndLU<'_>) -> Json<LdapConfigOut<'_>> {
    Json(cfg_and_lu.cfg.ldap.to_js_ui())
}
#[get("/config/remotes")]
fn config_remotes(cfg_and_lu : CfgAndLU<'_>) -> Json<&BTreeMap<String, RemoteConfig>> {
    Json(&cfg_and_lu.cfg.remotes)
}

pub fn routes() -> Vec<Route> {
    routes![
        login,
        clear_cache,
        clear_test_data, add_test_data, set_test_data, 
        sgroup, sgroup_direct_rights, group_flattened_mright, search_sgroups, search_subjects, mygroups,
        config_public, config_subject_sources, config_remotes,
        create, modify_sgroup_attrs, delete, modify_members_or_rights,
    ]
}

File: server/api_routes.ts
import express from 'express';
import { throw_ } from './helpers';
import * as cas_auth from './cas_auth'
import * as test_data from './test_data'
import * as api_get from './api_get'
import * as api_post from './api_post'
import * as cache from './cache'
import conf from './conf';
import { hLdapConfig, MonoAttrs, MyMods, RemoteSqlQuery, toDn } from './my_types';
import { query_params, q, orig_url, logged_user, query_opt_params, handleJsonP, handleVoidP, handleJson } from './express_helpers';

const api = express.Router();

// eslint-disable-next-line @typescript-eslint/no-misused-promises
api.get("/login", async (req, res) => {
    try {
        const { target, ticket } = query_params(req, { target: q.string, ticket: q.string })
        if (!target.startsWith('/') || target.startsWith("//")) {
            throw `invalid target ${target}, it must be a path-absolute url`
        }
        const service = orig_url(req).match(/(.*)ticket=/)?.[1] ?? throw_("weird login url");
        const user = await cas_auth.validate_ticket(conf.cas.prefix_url, service, ticket)
        // @ts-expect-error (req.session is not typed)
        req.session.user = user;
        res.redirect(target)
    } catch (e) {
        console.error(e)
        res.status(500);
        res.send("internal error")
    }
})

api.get("/set_test_data", handleVoidP(test_data.set))
api.get("/clear_test_data", handleVoidP(test_data.clear))
api.get("/add_test_data", handleVoidP(test_data.add))

api.post("/create", handleVoidP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    await api_post.create(logged_user(req), id, req.body as MonoAttrs)
}))

api.post("/modify_sgroup_attrs", handleVoidP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    await api_post.modify_sgroup_attrs(logged_user(req), id, req.body as MonoAttrs)
}))

api.post("/delete", handleVoidP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    await api_post.delete_(logged_user(req), id)
}))

api.post("/modify_members_or_rights", handleVoidP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    const { msg } = query_opt_params(req, { msg: q.string })
    await api_post.modify_members_or_rights(logged_user(req), id, req.body as MyMods, msg)
}))

api.post("/modify_remote_sql_query", handleVoidP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    const { msg } = query_opt_params(req, { msg: q.string })
    await api_post.modify_remote_sql_query(logged_user(req), id, req.body as RemoteSqlQuery, msg)
}))

api.get("/test_remote_query_sql", handleJsonP(async (req) => {
    const { id, remote_sql_query } = query_params(req, { id: q.string, remote_sql_query: q.json<RemoteSqlQuery>() })
    return await api_get.test_remote_query_sql(logged_user(req), id, remote_sql_query)
}))

api.get("/sgroup", handleJsonP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    return await api_get.get_sgroup(logged_user(req), id)
}))

api.get("/sgroup_direct_rights", handleJsonP(async (req) => {
    const { id } = query_params(req, { id: q.string })
    return await api_get.get_sgroup_direct_rights(logged_user(req), id)
}))

api.get("/group_flattened_mright", handleJsonP(async (req) => {
    const { id, mright } = query_params(req, { id: q.string, mright: q.mright })
    const { search_token, sizelimit } = query_opt_params(req, { search_token: q.string, sizelimit: q.int })
    return await api_get.get_group_flattened_mright(logged_user(req), id, mright, search_token, sizelimit)
}))

api.get("/sgroup_logs", handleJsonP(async (req) => {
    const { id, bytes } = query_params(req, { id: q.string, bytes: q.int })
    return await api_get.get_sgroup_logs(logged_user(req), id, bytes)
}))

api.get("/search_sgroups", handleJsonP(async (req) => {
    const { right, search_token, sizelimit } = query_params(req, { right: q.right, search_token: q.string, sizelimit: q.int })
    return await api_get.search_sgroups(logged_user(req), right, search_token, sizelimit)
}))

api.get("/mygroups", handleJsonP(async (req) => {
    return await api_get.mygroups(logged_user(req))
}))

api.get("/clear_cache", () => {
    cache.clear_all();
})

api.get("/search_subjects", handleJsonP(async (req) => {
    const { search_token, sizelimit } = query_params(req, { search_token: q.string, sizelimit: q.int })
    const { source_dn } = query_opt_params(req, { source_dn: q.string })
    return await api_get.api_search_subjects(logged_user(req), search_token, sizelimit, source_dn?.oMap(toDn))
}))

api.get("/config/public", handleJson(() => ({ "cas_prefix_url": conf.cas.prefix_url })))
api.get("/config/ldap", handleJson(() => hLdapConfig.to_js_ui(conf.ldap)))
api.get("/config/remotes", handleJson(() => conf.remotes))

export default api
